reps/2023-02-01-wasm-on-ray.md

```diff
@@ -0,0 +1,253 @@
+## Summary
+
+### General Motivation
+
+Currently, the supported workers in Ray are Python, Java, and C++. However, there are many other languages that are popular and have a large user base. For example, JavaScript, Go, and Rust, etc. It would be great if we can support these languages as well. However, adding support for these languages is not trivial. First of all, the language specific workers need to be supported in Ray core. Secondly, the relevant language specific libraries need to be developed such as Ray AIR.
+
+To make it easier to support new languages, we can use WebAssembly (WASM) to support new languages. WASM is a portable binary instruction format for a stack-based virtual machine. It is designed as a compilation target and can run with a minimum amount of overhead. There is an increasing number of languages that can be compiled to WASM, such as C/C++, Rust, Go. In addition, with the embedded WASM interpreter, we can support languages that cannot be compiled to WASM, such as JavaScript, Python, etc.
+
+Furthermore, it is also possible that we enable transparent Ray support in WASM. As a result, many legacy applications can be easily migrated to Ray and run in a distributed manner and Ray layer can be transparent to the application. This will be a huge benefit for Ray adoption in the generic serverless computing environment.
+
+In our internal FaaS environment, we have WASM function support. These functions are compiled from different programming languages such as Javascript, Rust, Go, C/C++, etc. Compared with normal container based FaaS solutions such as Knative, WASM based functions can be loaded with a shorter period of time while still providing the same or even better secure running environments. So far, WASM functions are still far from perfect. There are few pain points for WASM functions.
```

Comment:
This seems like a big adds to WASM if Ray can have WASM workers, in summary, seems like:

Ray adds on WASM:
- ObjectRef for communication
- Ray worker for parallelism and scale up

WASM adds on Ray:
- Lightweight x-lang support
- Lightweight isolation

Reply:
cc @scv119